feat: add `blas/ext/base/wasm/dnansumkbn2`

ShabiShett07 opened the pull request:
---
type: pre_push_report
description: Results of running various checks prior to pushing changes. report:
  - task: run_javascript_examples status: na
  - task: run_c_examples status: na
  - task: run_cpp_examples status: na
  - task: run_javascript_readme_examples status: na
  - task: run_c_benchmarks status: na
  - task: run_cpp_benchmarks status: na
  - task: run_fortran_benchmarks status: na
  - task: run_javascript_benchmarks status: na
  - task: run_julia_benchmarks status: na
  - task: run_python_benchmarks status: na
  - task: run_r_benchmarks status: na
  - task: run_javascript_tests status: na ---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes. report:
  - task: run_javascript_examples status: na
  - task: run_c_examples status: na
  - task: run_cpp_examples status: na
  - task: run_javascript_readme_examples status: na
  - task: run_c_benchmarks status: na
  - task: run_cpp_benchmarks status: na
  - task: run_fortran_benchmarks status: na
  - task: run_javascript_benchmarks status: na
  - task: run_julia_benchmarks status: na
  - task: run_python_benchmarks status: na
  - task: run_r_benchmarks status: na
  - task: run_javascript_tests status: na ---

Resolves #5732 

## Description

> What is the purpose of this pull request?

This pull request:

-   Solves #5732 

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #5732  

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

ShabiShett07 committed:
feat: add /blas/ext/base/wasm/dnansumkbn2

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

ShabiShett07 committed:
docs: new files

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: failed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: failed
---

ShabiShett07 committed:
docs: new files

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

ShabiShett07 committed:
fix: update benchmark files

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

ShabiShett07 committed:
docs: add test files

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

ShabiShett07 committed:
fix: package.json errors

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

ShabiShett07 committed:
fix: package.json errors

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

ShabiShett07 committed:
fix: package.json errors

0PrashantYadav0 commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/README.md:
```diff
@@ -0,0 +1,293 @@
+## Notes
+
+-   If `N <= 0`, both `main` and `ndarray` methods return `0.0`.
+-   This package implements routines using WebAssembly. When provided arrays which are not allocated on a `dnansumkbn2` module memory instance, data must be explicitly copied to module memory prior to computation. Data movement may entail a performance cost, and, thus, if you are using arrays external to module memory, you should prefer using [`@stdlib/blas/base/dnansumkbn2`][@stdlib/blas/ext/bas ...
```
```suggestion
-   This package implements routines using WebAssembly. When provided arrays which are not allocated on a `dnansumkbn2` module memory instance, data must be explicitly copied to module memory prior to computation. Data movement may entail a performance cost, and, thus, if you are using arrays external to module memory, you should prefer using [`@stdlib/blas/ext/base/dnansumkbn2`][@stdlib/blas/ext/base/dnansumkbn2]. However, if working with arrays which are allocated and explicitly managed on module memory, you can achieve better performance when compared to the pure JavaScript implementations found in [`@stdlib/blas/ext/base/dnansumkbn2`][@stdlib/blas/ext/base/dnansumkbn2]. Beware that such performance gains may come at the cost of additional complexity when having to perform manual memory management. Choosing between implementations depends heavily on the particular needs and constraints of your application, with no one choice universally better than the other.
```

ShabiShett07 committed:
fix: add appropriate links

Co-authored-by: JoyBoy <144602492+0PrashantYadav0@users.noreply.github.com>
Signed-off-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>

ShabiShett07 commented:
@0PrashantYadav0, may I know what was the changes needed because both the things look similar

0PrashantYadav0 commented:
In the docs you have ref to extend blas routines `dnansumkbn2` but in preview it shows blas routines `dnansumkbn2` which is not correct.

0PrashantYadav0 commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/lib/routine.js:
```diff
@@ -0,0 +1,166 @@
+});
+
+/**
+* Adds a scalar constant to each double-precision floating-point strided array element and computes the sum using an improved Kahan–Babuška algorithm and alternative indexing semantics.
```
```suggestion
* Computes the sum of double-precision floating-point strided array elements, ignoring `NaN` values and using a second-order iterative Kahan–Babuška algorithm and alternative indexing semantics.
```

0PrashantYadav0 commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/package.json:
```diff
@@ -0,0 +1,79 @@
+    "win32",
+    "windows"
+  ],
+  "keywords": [
```
add keywords like `"webassembly"`, `"wasm"`

0PrashantYadav0 commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/test/test.ndarray.js:
```diff
@@ -0,0 +1,206 @@
+
+var tape = require( 'tape' );
+var Float64Array = require( '@stdlib/array/float64' );
+var dasumpw = require( './../lib' );
```
you need to change the test here also change `dasumpw` to `dnansumkbn2`

0PrashantYadav0 commented:
Comment : Please check your tests again.

ShabiShett07 committed:
chore: change statements

Co-authored-by: JoyBoy <144602492+0PrashantYadav0@users.noreply.github.com>
Signed-off-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>

ShabiShett07 committed:
chore: add keywords

Signed-off-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>

ShabiShett07 committed:
fix: update variables

Signed-off-by: Shabareesh Shetty <139731143+ShabiShett07@users.noreply.github.com>

ShabiShett07 commented:
@0PrashantYadav0, thank you for the corrections, I have made changes according to your suggestions

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into feature/wasm/dnansumkbn2

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/docs/repl.txt:
```diff
@@ -0,0 +1,501 @@
+    > var xptr = 0;
+
+    // Write data to module memory:
+    > mod.write( xptr, [ 1, 2, 3, NaN ] );
```
```suggestion
    > mod.write( xptr, new {{alias:@stdlib/array/float64}}( [ 1.0, 2.0, 3.0, NaN ] ) );
```

You need to be explicit regarding the dtype when using `write`.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/docs/repl.txt:
```diff
@@ -0,0 +1,501 @@
+    > var xptr = 0;
+
+    // Write data to module memory:
+    > mod.write( xptr, [ 1, 2, 3, NaN ] );
```
```suggestion
    > mod.write( xptr, new {{alias:@stdlib/array/float64}}( [ 1.0, 2.0, 3.0, NaN ] ) );
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/docs/types/index.d.ts:
```diff
@@ -0,0 +1,313 @@
+	* var xptr = 0;
+	*
+	* // Write vector values to module memory:
+	* mod.write( xptr, [ 1, 2, 3, NaN] );
```
Same thing here and everywhere. While it is not strictly necessary here because of the default behavior of `write`, you'll save everyone copy-paste errors if you are explicit regarding the array type passed to `write`. Use `array/float64` to pass in a `Float64Array`. Also, your spacing is inconsistent.

You should apply the same changes here and elsewhere throughout this PR.

0PrashantYadav0 replied:
i am using `mod.write( xptr, oneTo( N, dtype ) );` in my pr instead of what you mentioned above.
Here in this file dtype and N is declared but never used.

And for repl.txt i am using `{{alias:@stdlib/array/one-to}}( 3, 'float64' )`

0PrashantYadav0 replied:
@kgryte If there are changes in above please do let me know as i have to make those changes in my pr as well

kgryte replied:
@0PrashantYadav0  `{{alias:@stdlib/array/one-to}}( 3, 'float64' )` is fine to use and the reason is can be ascertained by the use of `'float64'` in that expression. Namely, `@stdlib/array/one-to` is invoked to return a `Float64Array`. Here, @ShabiShett07 didn't want to use `one-to` because he wanted the example to include `NaN` values. That is fine, but, if he is going to do that, he needs to wrap in a `Float64Array( [...] )`.

0PrashantYadav0 replied:
Okay that make sense...
Just one more question do we have to include similar changes to other wasm packages ?

kgryte replied:
@0PrashantYadav0 Only if you are wanting the examples to show a `NaN` example. I am fine with the examples using `one-to`, so, if you don't need to update the examples, then no work is necessary. The only instance where you want to update the examples is when a provided array is not typed. Meaning, don't pass a generic array.

0PrashantYadav0 replied:
Ok thank you for update

kgryte reviewed (changes requested):
Left initial comments.

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into feature/wasm/dnansumkbn2

ShabiShett07 committed:
chore: add array datatype

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 commented:
@kgryte, I have added Float64Array in both index.d.ts and index.js

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/docs/types/index.d.ts:
```diff
@@ -0,0 +1,320 @@
+	*
+	* // Write vector values to module memory:
+	* var x = new Float64Array( [ 1, 2, 3, NaN ] );
+	*
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/docs/types/index.d.ts:
```diff
@@ -0,0 +1,320 @@
+	*
+	* // Write vector values to module memory:
+	* var x = new Float64Array( [ 1, 2, 3, NaN ] );
+	*
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/docs/types/index.d.ts:
```diff
@@ -0,0 +1,320 @@
+	*
+	* // Write vector values to module memory:
+	* var x = new Float64Array( [ 1, 2, 3, NaN ] );
+	*
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/lib/module.js:
```diff
@@ -0,0 +1,197 @@
+* var xptr = 0;
+*
+* // Write vector values to module memory:
+* dnansumkbn2.write( xptr, [ 1, 2, 3, NaN ] );
```
@ShabiShett07 You didn't update these examples. I suggest doing another review of this PR to ensure you have updated all the necessary examples.

ShabiShett07 committed:
chore: add array datatype

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 commented:
@kgryte, I have made the required changes.

kgryte committed:
docs: fix missing imports

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
docs: fix missing decimals

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/docs/types/index.d.ts:
```diff
@@ -0,0 +1,319 @@
+	*
+	* // Define a vector data type:
+	* var dtype = 'float64';
+	*
```
```suggestion
```

@ShabiShett07 If you are going to change examples, you need to be sure that you edit the entire example to ensure that it makes sense. When we used `oneTo`, we needed a `dtype`, but that is no longer the case, as you are wanting to use `Float64Array`.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/docs/types/index.d.ts:
```diff
@@ -0,0 +1,319 @@
+	* // Initialize the routine:
+	* mod.initializeSync();
+	*
+	* // Define a vector data type:
+	* var dtype = 'float64';
+	*
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/docs/types/index.d.ts:
```diff
@@ -0,0 +1,319 @@
+	*
+	* // Define a vector data type:
+	* var dtype = 'float64';
+	*
```
```suggestion
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/docs/types/test.ts:
```diff
@@ -0,0 +1,345 @@
+	dnansumkbn2.main( ( x: number ): number => x, x, 1 ); // $ExpectError
+}
+
+// The compiler throws an error if the `main` method is provided a second argument which is not a number...
```
```suggestion
// The compiler throws an error if the `main` method is provided a second argument which is not a Float64Array...
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/docs/types/test.ts:
```diff
@@ -0,0 +1,345 @@
+	dnansumkbn2.main( x.length, ( x: number ): number => x, 1 ); // $ExpectError
+}
+
+// The compiler throws an error if the `main` method is provided a third argument which is not a Float64Array...
```
```suggestion
// The compiler throws an error if the `main` method is provided a third argument which is not a number...
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/docs/types/test.ts:
```diff
@@ -0,0 +1,345 @@
+	dnansumkbn2.ndarray( ( x: number ): number => x, x, 1, 0 ); // $ExpectError
+}
+
+// The compiler throws an error if the `ndarray` method is provided a second argument which is not a number...
```
```suggestion
// The compiler throws an error if the `ndarray` method is provided a second argument which is not a Float64Array...
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/docs/types/test.ts:
```diff
@@ -0,0 +1,345 @@
+	dnansumkbn2.ndarray( x.length, ( x: number ): number => x, 1, 0 ); // $ExpectError
+}
+
+// The compiler throws an error if the `ndarray` method is provided a third argument which is not a Float64Array...
```
```suggestion
// The compiler throws an error if the `ndarray` method is provided a third argument which is not a number...
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/examples/module.js:
```diff
@@ -0,0 +1,59 @@
+	var xptr = 0;
+
+	// Write vector values to module memory:
+	mod.write( xptr, [ 1, 2, 3, NaN ] );
```
@ShabiShett07 Another example where you need to use a `Float64Array`. You are also missing decimals.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/lib/index.js:
```diff
@@ -0,0 +1,96 @@
+* var xptr = 0;
+*
+* // Write vector values to module memory:
+* var x = new Float64Array( [ 1, 2, 3, NaN ] );
```
Missing decimals.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/lib/main.js:
```diff
@@ -0,0 +1,62 @@
+* var x = new Float64Array( [ 2.0, 1.0, NaN, -2.0, -2.0, 2.0, 3.0, 4.0 ] );
+*
+* // Perform operation:
+* var v = dnansumkbn2.ndarray( 4, x, 2, 1 );
```
As a heads up: because of the offset value, `NaN` is not actually included in the list of iterated elements; it is skipped over. Is that intentional?

ShabiShett07 replied:
No, that was just just a mistake, I will shift it to the next index

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/lib/module.js:
```diff
@@ -0,0 +1,200 @@
+*
+* @example
+* var Memory = require( '@stdlib/wasm/memory' );
+* var oneTo = require( '@stdlib/array/one-to' );
```
If you are not going to use this, you need to update to import the `Float64Array` constructor.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/lib/module.js:
```diff
@@ -0,0 +1,200 @@
+* dnansumkbn2.initializeSync();
+*
+* // Define a vector data type:
+* var dtype = 'float64';
```
This isn't used. All of these comments likely apply to your other examples. I suggest, once again, going back through this PR with a fine tooth comb looking for various mistakes such as these.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/lib/module.js:
```diff
@@ -0,0 +1,200 @@
+* var xptr = 0;
+*
+* // Write vector values to module memory:
+* var x = new Float64Array( [ 1, 2, 3, NaN ] );
```
You're missing decimals.

kgryte replied:
It is a double-precision floating-point array. Always include decimals for float values.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/lib/module.js:
```diff
@@ -0,0 +1,200 @@
+* var xptr = 0;
+*
+* // Write vector values to module memory:
+* var x = new Float64Array( [ 1, 2, 3, NaN ] );
```
Same comments. You are also missing the `Float64Array` import.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/lib/module.js:
```diff
@@ -0,0 +1,200 @@
+* dnansumkbn2.initializeSync();
+*
+* // Define a vector data type:
+* var dtype = 'float64';
```
Not used.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/lib/module.js:
```diff
@@ -0,0 +1,200 @@
+*
+* @example
+* var Memory = require( '@stdlib/wasm/memory' );
+* var oneTo = require( '@stdlib/array/one-to' );
```
All the same comments.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/lib/routine.js:
```diff
@@ -0,0 +1,166 @@
+* dnansumkbn2.initializeSync();
+*
+* // Define a strided array:
+* var x = new Float64Array( [ 2.0, 1.0, NaN, -2.0, -2.0, 2.0, 3.0, 4.0 ] );
```
Same comment. You skip the `NaN` here. Intentional?

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/lib/routine.js:
```diff
@@ -0,0 +1,166 @@
+* var x = new Float64Array( [ 2.0, 1.0, NaN, -2.0, -2.0, 2.0, 3.0, 4.0 ] );
+*
+* // Perform operation:
+* var sum = dnansumkbn2.ndarray( 4, x, 2, 1 );
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/package.json:
```diff
@@ -0,0 +1,81 @@
+{
+  "name": "@stdlib/blas/ext/base/wasm/dnansumkbn2",
+  "version": "0.0.0",
+  "description": "Computes the sum of double-precision floating-point strided array elements, ignoring `NaN` values and using a second-order iterative Kahan–Babuška algorithm.",
```
```suggestion
  "description": "Compute the sum of double-precision floating-point strided array elements, ignoring `NaN` values and using a second-order iterative Kahan–Babuška algorithm.",
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/test/test.main.js:
```diff
@@ -0,0 +1,191 @@
+	t.end();
+});
+
+tape( 'the function calculates the sum of absolute values', function test( t ) {
```
Description is wrong.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/test/test.main.js:
```diff
@@ -0,0 +1,191 @@
+		2.0,
+		2.0,  // 1
+		-7.0,
+		NaN, // 2
```
```suggestion
		NaN,  // 2
```

Alignment. Attention to detail.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/test/test.main.js:
```diff
@@ -0,0 +1,191 @@
+		2.0,
+		2.0,  // 2
+		-7.0,
+		NaN, // 1
```
```suggestion
		NaN,  // 1
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/test/test.main.js:
```diff
@@ -0,0 +1,191 @@
+		2.0,
+		1.0,  // 0
+		2.0,
+		NaN, // 1
```
```suggestion
		NaN,  // 1
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/test/test.module.main.js:
```diff
@@ -0,0 +1,202 @@
+	t.end();
+});
+
+tape( 'a module instance has a `main` method that the sum of absolute values of all strided array elements', function test( t ) {
```
Description is wrong.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/test/test.module.main.js:
```diff
@@ -0,0 +1,202 @@
+	t.end();
+});
+
+tape( 'if provided an `N` parameter equal to `1`, a module instance has a `main` method which returns the absolute value of the first element', function test( t ) {
```
Description is wrong. Please be careful of copy-paste errors. I suggest doing a close look at `blas/ext/base/dnansumkbn2`.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/test/test.module.main.js:
```diff
@@ -0,0 +1,202 @@
+		2.0,
+		2.0,  // 1
+		-7.0,
+		NaN, // 2
```
Alignment here and everywhere else.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/test/test.module.ndarray.js:
```diff
@@ -0,0 +1,233 @@
+	t.end();
+});
+
+tape( 'a module instance has a `ndarray` method that the sum of absolute values of all strided array elements', function test( t ) {
```
Description.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/test/test.module.ndarray.js:
```diff
@@ -0,0 +1,233 @@
+
+	t.end();
+});
+
```
See `blas/ext/base/dnansumkbn2`. A new set of tests has been added for signed zero.

kgryte replied:
Note that you may need to also regenerate the Wasm binary, as the upstream implementation has changed.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/test/test.module.ndarray.js:
```diff
@@ -0,0 +1,233 @@
+	t.end();
+});
+
+tape( 'if provided an `N` parameter equal to `1`, a module instance has a `ndarray` method which returns the absolute value of the first element', function test( t ) {
```
Description.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/test/test.module.ndarray.js:
```diff
@@ -0,0 +1,233 @@
+	t.end();
+});
+
+tape( 'if provided an `stride` parameter equal to `0`, a module instance has a `ndarray` method which returns the absolute value of the first element repeated N times', function test( t ) {
```
Description.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/test/test.ndarray.js:
```diff
@@ -0,0 +1,206 @@
+	t.end();
+});
+
+tape( 'the function calculates the sum of absolute values', function test( t ) {
```
Descriptions here and elsewhere.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/test/test.ndarray.js:
```diff
@@ -0,0 +1,206 @@
+
+	t.end();
+});
+
```
Add signed zero test.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/test/test.ndarray.js:
```diff
@@ -0,0 +1,206 @@
+		2.0,
+		2.0,  // 1
+		-7.0,
+		NaN, // 2
```
Alignment.

kgryte reviewed (changes requested):
Left another round of comments. As suggested previously, please do a close review of the files in this PR. There are still a number of mistakes that need cleaning up.

ShabiShett07 committed:
chore: update implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

ShabiShett07 committed:
chore: update implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into feature/wasm/dnansumkbn2

ShabiShett07 committed:
chore: update implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

ShabiShett07 commented:
@kgryte, I have tried to find all mistakes and resolved them. Also I have updated the implementation according to the updated package

0PrashantYadav0 commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/README.md:
```diff
@@ -0,0 +1,293 @@
+mod.initializeSync();
+
+// Define a vector data type:
+var dtype = 'float64';
```
Remove these if you are not using this

0PrashantYadav0 commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/README.md:
```diff
@@ -0,0 +1,293 @@
+mod.initializeSync();
+
+// Define a vector data type:
+var dtype = 'float64';
```
Remove if you are not using this

0PrashantYadav0 commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/README.md:
```diff
@@ -0,0 +1,293 @@
+var xptr = 0;
+
+// Write vector values to module memory:
+mod.write( xptr, [ 1, 2, 3, NaN ] );
```
```suggestion
mod.write( xptr, [ 1.0, 2.0, 3.0, NaN ] );
```

0PrashantYadav0 commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/README.md:
```diff
@@ -0,0 +1,293 @@
+var xptr = 0;
+
+// Write vector values to module memory:
+mod.write( xptr, [ 1, 2, 3, NaN ] );
```
```suggestion
mod.write( xptr, [ 1.0, 2.0, 3.0, NaN ] );
```

0PrashantYadav0 commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/benchmark/benchmark.js:
```diff
@@ -0,0 +1,118 @@
+* @private
+* @returns {number} random number
+*/
+function rand() {
```
This function works fine but we have different way to expressing the `rand()` function. Refer the other packages.
Apply to all the instances.

ShabiShett07 replied:
@0PrashantYadav0, thanks a lot for the review. I think this is intentional, so that we can check whether the condition occurs when a `NaN` value is passed to the package. Does it make sense?

kgryte replied:
Yes, this looks fine to me. @0PrashantYadav0 what is the change you were looking for explicitly?

0PrashantYadav0 replied:
Yes it is fine by me, but usually other all other packages have different way of describing `rand` function.

kgryte replied:
@0PrashantYadav0 If you are referring to "random number or NaN", that description is actually **not** the desired description as that description is too literal. What is used here is the preferred description.

ShabiShett07 committed:
chore: add decimal values

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/README.md:
```diff
@@ -0,0 +1,287 @@
+var xptr = 0;
+
+// Write vector values to module memory:
+mod.write( xptr, [ 1.0, 2.0, 3.0, NaN ] );
```
@ShabiShett07 This needs to be a `Float64Array`.

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/README.md:
```diff
@@ -0,0 +1,287 @@
+var xptr = 0;
+
+// Write vector values to module memory:
+mod.write( xptr, [ 1.0, 2.0, 3.0, NaN ] );
```
@ShabiShett07 This needs to be a `Float64Array`.

kgryte committed:
docs: fix descriptions

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/lib/module.js:
```diff
@@ -0,0 +1,194 @@
+*
+* @example
+* var Memory = require( '@stdlib/wasm/memory' );
+* var oneTo = require( '@stdlib/array/one-to' );
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/ext/base/wasm/dnansumkbn2/lib/module.js:
```diff
@@ -0,0 +1,194 @@
+*
+* @example
+* var Memory = require( '@stdlib/wasm/memory' );
+* var oneTo = require( '@stdlib/array/one-to' );
```
```suggestion
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
docs: use `Float64Array`

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (approved):
LGTM after clean-up.
